gk/fib.c

```diff
@@ -0,0 +1,1440 @@
+				ip_prefix->str);
+			return NULL;
+		}
+	}
+
+	return ip_prefix_fib;
+}
+
+static int
+ether_cache_put(struct gk_fib *neighbor_fib,
+	struct ether_cache *eth_cache, struct ipaddr *addr)
+{
+	if (eth_cache->ref_cnt == 0) {
+		rte_panic("Unexpected condition: the ref_cnt of the ether cache is 0 at %s\n",
+			__func__);
+	}
+
+	if (eth_cache->ref_cnt > 1) {
+		eth_cache->ref_cnt--;
+		return 0;
+	}
+
+	if (addr->proto == ETHER_TYPE_IPv4) {
+		return neigh_del_ether_cache(
+			&neighbor_fib->u.neigh, &addr->ip.v4.s_addr);
+	}
+
+	if (addr->proto == ETHER_TYPE_IPv6) {
+		return neigh_del_ether_cache(
+			&neighbor_fib->u.neigh6, addr->ip.v6.s6_addr);
+	}
+
+	RTE_LOG(ERR, GATEKEEPER,
+		"gk: remove an invalid FIB entry with IP type %hu at %s",
+		addr->proto, __func__);
+
+	return -1;
+}
+
+/*
+ * Note that, @action should be either GK_FWD_GATEWAY_FRONT_NET
+ * or GK_FWD_GATEWAY_BACK_NET.
+ */
+static struct gk_fib *
+find_fib_entry_for_neighbor_locked(
+	struct ipaddr *gw_addr, enum gk_fib_action action,
+	struct gatekeeper_if *iface, struct gk_config *gk_conf)
+{
+	int fib_id;
+	struct gk_fib *neigh_fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+
+	if (gw_addr->proto == ETHER_TYPE_IPv4 &&
+			ipv4_if_configured(iface)) {
+		fib_id = lpm_lookup_ipv4(ltbl->lpm, gw_addr->ip.v4.s_addr);
+		/*
+		 * Invalid gateway entry, since at least we should
+		 * obtain the FIB entry for the neighbor table.
+		 */
+		if (fib_id < 0)
+			return NULL;
+
+		neigh_fib = &ltbl->fib_tbl[fib_id];
+
+		/*
+		 * Invalid gateway entry, since the neighbor entry
+		 * and the gateway entry should be in the same network.
+		 */
+		if ((action == GK_FWD_GATEWAY_FRONT_NET &&
+				neigh_fib->action != GK_FWD_NEIGHBOR_FRONT_NET)
+				|| (action == GK_FWD_GATEWAY_BACK_NET &&
+				neigh_fib->action !=
+				GK_FWD_NEIGHBOR_BACK_NET)) {
+			return NULL;
+		}
+
+		return neigh_fib;
+	} else if (likely(gw_addr->proto == ETHER_TYPE_IPv6)
+			&& ipv6_if_configured(iface)) {
+		fib_id = lpm_lookup_ipv6(ltbl->lpm6, gw_addr->ip.v6.s6_addr);
+		/*
+		 * Invalid gateway entry, since at least we should
+		 * obtain the FIB entry for the neighbor table.
+		 */
+		if (fib_id < 0)
+			return NULL;
+
+		neigh_fib = &ltbl->fib_tbl6[fib_id];
+
+		/*
+		 * Invalid gateway entry, since the neighbor entry
+		 * and the gateway entry should be in the same network.
+		 */
+		if ((action == GK_FWD_GATEWAY_FRONT_NET &&
+				neigh_fib->action != GK_FWD_NEIGHBOR_FRONT_NET)
+				|| (action == GK_FWD_GATEWAY_BACK_NET &&
+				neigh_fib->action !=
+				GK_FWD_NEIGHBOR_BACK_NET)) {
+			return NULL;
+		}
+
+		return neigh_fib;
+	} else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: unconfigued IP type %hu at interface %s!\n",
+			gw_addr->proto, iface->name);
+	}
+
+	return NULL;
+}
+
+/*
+ * This function is called by del_fib_entry_locked().
+ * Notice that, it doesn't stand on its own, and it's only
+ * a construct to make del_fib_entry_locked() readable.
+ */
+static int
+del_gateway_from_neigh_table_locked(
```

Comment:
What this function really does is to look `neigh_fib` up so it can call `ether_cache_put()`. The problem with this is that `ether_cache_put()` doesn't alway use `neigh_fib`, but the lookup is always done.

A solution is to do the lookup in `ether_cache_put()` when needed. `ether_cache_put()` would still have parameter `neighbor_fib` but it could be `NULL`. Of course, the lookup would only be done when needed and `neighbor_fib` is `NULL`.

`ether_cache_put()` won't grow that much because other comments are going to shrink the body of `del_gateway_from_neigh_table_locked()`.